### PR TITLE
`reserve` micro optimisations for vectors/smallvectors and use correct type for vector size

### PR DESCRIPTION
**Pull request by SamuelMarks**

### CODE REVIEW

File: c10/cuda/CUDACachingAllocator.cpp [hunk 1]
```diff
@@ -1150,14 +1150,14 @@ static std::string reportProcessMemoryInfo(int device) {
           pci_id, &nvml_device));
 
   std::vector<nvmlProcessInfo_v1_t> procs(8);
-  unsigned int size = procs.size();
+  std::vector<decltype(nvmlProcessInfo_v1_t)>::size_type size = procs.size();
```
Comment (Skylion007) [inline c10/cuda/CUDACachingAllocator.cpp:1153]: Couldn't this just be auto?

**Review by Skylion007 — APPROVED** [review]

File: c10/cuda/CUDACachingAllocator.cpp (hunk 1 above)
Comment (SamuelMarks) [reply c10/cuda/CUDACachingAllocator.cpp:1153]: Yes that would work also. I prefer explicit over implicit but sure I'll commit this.

**Review by Skylion007 — APPROVED** [review]

File: c10/cuda/CUDACachingAllocator.cpp [hunk 2]
```diff
@@ -1150,14 +1150,14 @@ static std::string reportProcessMemoryInfo(int device) {
...
-  unsigned int size = procs.size();
+  auto size = procs.size();
   nvmlReturn_t r;
   while ((r = DriverAPI::get()->nvmlDeviceGetComputeRunningProcesses_(
               nvml_device, &size, procs.data())) ==
          NVML_ERROR_INSUFFICIENT_SIZE) {
     procs.resize(size);
   }
-  unsigned int self_pid = getpid();
+  pid_t self_pid = getpid();
```
Comment (Skylion007) [inline c10/cuda/CUDACachingAllocator.cpp:1160]: This is causing a build error because ti's being compared with unsigned int later in the method.

Comment (SamuelMarks) [reply c10/cuda/CUDACachingAllocator.cpp:1160]: Weird, `getpid` is `pid_t` everywhere:
- Linux: https://man7.org/linux/man-pages/man2/getpid.2.html https://linux.die.net/man/2/getpid
- POSIX: https://pubs.opengroup.org/onlinepubs/009696699/functions/getpid.html
- zOS: https://www.ibm.com/docs/en/zos/2.4.0?topic=functions-getpid-get-process-id
- FreeBSD: https://man.freebsd.org/cgi/man.cgi?query=getpid&sektion=2

File: c10/cuda/CUDACachingAllocator.cpp [hunk 3]
```diff
@@ -1150,7 +1150,7 @@ static std::string reportProcessMemoryInfo(int device) {
           pci_id, &nvml_device));
 
   std::vector<nvmlProcessInfo_v1_t> procs(8);
-  unsigned int size = procs.size();
+  auto size = procs.size();
```
Comment (Skylion007) [inline c10/cuda/CUDACachingAllocator.cpp:1153]: This implicit cache was necessary as removing the cast breaks the build it seems...

Comment (SamuelMarks) [reply c10/cuda/CUDACachingAllocator.cpp:1153]: Ok so back to `std::vector<decltype(nvmlProcessInfo_v1_t)>::size_type`?

Comment (Skylion007) [reply c10/cuda/CUDACachingAllocator.cpp:1153]: @SamuelMarks Did that actually not generate a compile error before? Does procs.size() not equal `std::vector<decltype(nvmlProcessInfo_v1_t)>::size_type`? It seems to on every platform except some versions of cuda? If so maybe make it clear with a static_cast / comment to explain why we are doing that. `auto size = static_cast<std::vector<decltype(nvmlProcessInfo_v1_t)>::size_type>(procs.size())` ?

Comment (SamuelMarks) [reply c10/cuda/CUDACachingAllocator.cpp]: Hmm, well it looks like this variant worked on CI `std::vector<decltype(procs)>::size_type size`

File: c10/cuda/CUDACachingAllocator.cpp [hunk 4]
```diff
@@ -1150,7 +1150,7 @@ static std::string reportProcessMemoryInfo(int device) {
           pci_id, &nvml_device));
 
   std::vector<nvmlProcessInfo_v1_t> procs(8);
-  unsigned int size = procs.size();
+  std::vector<decltype(procs)>::size_type size = procs.size();
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply c10/cuda/CUDACachingAllocator.cpp:1153]: Oh I thought you were saying that the `auto` didn't work; ok reverting

**Comment by SamuelMarks** [comment]
@Skylion007 Wait this wasn't merged? - Anything you need changed?

(I can merge in master)

**Comment by Skylion007** [comment]
> @Skylion007 Wait this wasn't merged? - Anything you need changed?
> 
> (I can merge in master)

Lintrunner needs to be fixed / tests need to pass.

File: c10/cuda/CUDACachingAllocator.cpp [hunk 5]
```diff
@@ -1205,7 +1205,7 @@ class DeviceCachingAllocator {
   // outstanding cuda events
   ska::flat_hash_map<
       cuda::CUDAStream,
-      std::deque<std::pair<EventPool::Event, Block*>>>
```
Comment (Skylion007) [inline c10/cuda/CUDACachingAllocator.cpp:1208]: f here is probably a typo

Comment (SamuelMarks) [reply c10/cuda/CUDACachingAllocator.cpp]: Ahh! - Thanks. Fixed and merged latest master

**Comment by Skylion007** [comment]
Build is still failing on a few targets

**Comment by SamuelMarks** [comment]
@Skylion007 Hmm, investigating:

#### [linux-bionic-cuda12.1-py3.10-gcc9-sm86](https://github.com/pytorch/pytorch/actions/runs/5780261825/job/15666090723?pr=102868#logs)
Issue @ c10/cuda/CUDACachingAllocator.cpp:1144 with this:
> Wno-invalid-partial-specialization

Slightly more interesting [linux-bionic-cuda12.1-py3.10-gcc9-sm86](https://github.com/pytorch/pytorch/actions/runs/5780261825/job/15666090723?pr=102868#logs):
> cannot convert ‘\u{1b}[01m\u{1b}[Klong unsigned int*\u{1b}[m\u{1b}[K’ to ‘\u{1b}[01m\u{1b}[Kunsigned int*\u{1b}[m\u{1b}[K’ in argument passing\n 1144 |               nvml_device, \u{1b}[01;31m\u{1b}[K&size\u{1b}[m\u{1b}[K, procs.data())) ==\n

So it's an `unsigned int*` and a `long unsigned int*` trying  to work with each other. Let me check it out

**Comment by SamuelMarks** [comment]
Ouch. Shall I open a new PR with the fix?

**Pull request closed without merging**